official-antistasi-community/A3-Antistasi
# CS-1489-Pun_FFServerSide

## What type of PR is this.
* Bug
* Change

### What have I changed and why?
* This change is required for future bug fixes.

* Punishment_FF moved serverside, this halves latency for a response from the server.
* Server functions directly call each other.
* All functions were looked at and received minor optimisations. Which probably is quite noticeable in the big picture.
* UID and player tracking backend was completely reworked.
  * Bonus: Fixes issue where testing FF with two local clients did not work as expected.
* Titles of some hint messages were changed to better integrate with new hint system.

**INCLUDES NestedObject PR Branch https://github.com/official-antistasi-community/A3-Antistasi/pull/1441**

### Please specify which Issue this PR Resolves.
closes https://github.com/official-antistasi-community/A3-Antistasi/issues/1489
closes https://github.com/official-antistasi-community/A3-Antistasi/issues/1477
closes https://github.com/official-antistasi-community/A3-Antistasi/issues/1767

### Please verify the following and ensure all checks are completed.

* [ ] Have you loaded the mission on a dedicated server with another person?

### Is further testing or are further changes required?
1. [x] No The Event Team has done extensive testing and the system has worked flawlessly.
* Yes

## Testing Steps

Rebel damage itself=> nothing
Rebel  in vehicle damage itself  in vehicle=> nothing  // Grenade/Console recommended  // See further below

Rebel damage Rebel => Punish
Rebel damage Rebel in vehicle => Punish
Rebel in vehicle damage Rebel => Punish
Rebel in vehicle damage Rebel in different vehicle => Punish


Rebel in vehicle damage Rebel in same vehicle => nothing  // Note: you will need to use grenades/explosives or console.
```sqf
// Look at attacker
John = cursorObject;
// Look at victim
Smith = cursorObject;
// Let them climb into vehicle, tem exec
[John, 10, 1, Smith, "No one should see this message."] remoteExec ["A3A_fnc_punishment_FF",John,false];
```

Engagements between rebels and Pvp should not cause FF.

********************************************************
